pkg/transport/generic_transport_controller.go

```diff
@@ -0,0 +1,512 @@
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create k8s clientset for given config - %w", err)
+	}
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(cacheddiscovery.NewMemCacheClient(clientset.Discovery()))
+
+	restMapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to discover GroupVersionResource from given GroupVersionKind - %w", err)
+	}
+
+	return &(restMapping.Resource), nil
+}
+
+type genericTransportController struct {
+	logger logr.Logger
+
+	placementDecisionLister         edgev1alpha1listers.PlacementDecisionLister
+	placementDecisionInformerSynced cache.InformerSynced
+	wrappedObjectInformerSynced     cache.InformerSynced
+
+	// workqueue is a rate limited work queue.
+	// This is used to queue work to be processed instead of performing it as soon as a change happens.
+	// This means we can ensure we only process a fixed amount of resources at a time, and makes it
+	// easy to ensure we are never processing the same item simultaneously in two different workers.
+	workqueue workqueue.RateLimitingInterface
+
+	//transport is a specific implementation for the transport interface.
+	transport        Transport
+	transportClient  dynamic.Interface // dynamic client to transport wrapped object. since object kind is unknown during complilation, we use dynamic
+	wrappedObjectGVR schema.GroupVersionResource
+
+	wdsClientset     *edgeclientset.Clientset
+	wdsDynamicClient *dynamic.DynamicClient
+	wdsName          string
+}
+
+// enqueuePlacementDecision takes an PlacementDecision resource and
+// converts it into a namespace/name string which is put onto the workqueue.
+// This func *shouldn't* handle any resource other than PlacementDecision.
+func (c *genericTransportController) enqueuePlacementDecision(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		c.logger.Error(err, "failed to enqueue PlacementDecision object")
+		return
+	}
+
+	c.workqueue.Add(key)
+}
+
+// handleWrappedObject takes transport-specific wrapped object resource,
+// extracts the origin PlacementDecision of the given wrapped object and
+// enqueue that PlacementDecision object for processing. This way, we
+// don't need to implement custom logic for handling WrappedObject resources.
+// More info on this pattern here:
+// https://github.com/kubernetes/community/blob/8cafef897a22026d42f5e5bb3f104febe7e29830/contributors/devel/controllers.md
+func (c *genericTransportController) handleWrappedObject(obj interface{}) {
+	wrappedObject := obj.(metav1.Object)
+	ownerPlacementDecisionKey, found := wrappedObject.GetAnnotations()[originOwnerReferenceAnnotation] // safe if GetAnnotations() returns nil
+	if !found {
+		c.logger.Info("failed to extract placementdecision key from transport wrapped object", "WrappedObjectName", wrappedObject.GetName())
```

Comment:
The namespace should also be logged, right?

Reply:
yes, missed it. 
done.